src/main/java/slimeknights/tconstruct/smeltery/tileentity/AbstractCastingTileEntity.java

```diff
@@ -0,0 +1,234 @@
+      // we therefore spill the whole contents on extraction.
+      ItemStack stack = getStackInSlot(slot);
+      if (slot == OUTPUT) {
+        // fire player smelt event?
+      }
+      ItemHandlerHelper.giveItemToPlayer(player, stack);
+      setInventorySlotContents(slot, ItemStack.EMPTY);
+
+      // send a block update for the comparator, needs to be done after the stack is removed
+      if (slot == OUTPUT) {
+        this.getWorld().notifyNeighborsOfStateChange(this.pos, this.getBlockState().getBlock());
+      }
+    }
+  }
+
+
+  @Override
+  @Nonnull
+  public int[] getSlotsForFace(Direction side) {
+    return new int[]{INPUT, OUTPUT};
+  }
+
+  @Override
+  public boolean canInsertItem(int index, ItemStack itemStackIn, @Nullable Direction direction) {
+    return index == INPUT && !isStackInSlot(OUTPUT);
+  }
+
+  @Override
+  public boolean canExtractItem(int index, ItemStack stack, Direction direction) {
+    return index == OUTPUT;
+  }
+
+  @Override
+  public void tick() {
+    // no recipe
+    if (recipe == null) {
+      return;
+    }
+    // fully filled
+    if (tank.getFluidAmount() == tank.getCapacity() && !tank.getFluid().isEmpty()) {
+      timer++;
+      if (!getWorld().isRemote) {
+        if (timer >= recipe.getCoolingTime()) {
+          if (recipe.isConsumed()) {
+            setInventorySlotContents(INPUT, ItemStack.EMPTY);
+            // Fire particles
+          }
+          if (recipe.switchSlots()) {
+            setInventorySlotContents(INPUT, recipe.getCraftingResult(crafting));
+          }
+          else {
+            setInventorySlotContents(OUTPUT, recipe.getCraftingResult(crafting));
+          }
+          getWorld().playSound(null, pos, SoundEvents.BLOCK_LAVA_EXTINGUISH, SoundCategory.AMBIENT, 0.07f, 4f);
+
+          reset();
+
+          getWorld().notifyNeighborsOfStateChange(this.pos, this.getBlockState().getBlock());
+        }
+      }
+      else if (getWorld().rand.nextFloat() > 0.9f) {
+        world.addParticle(ParticleTypes.SMOKE, pos.getX() + getWorld().rand.nextDouble(), pos.getY() + 1.1d, pos.getZ() + getWorld().rand.nextDouble(), 0.0D, 0.0D, 0.0D);
+      }
+    }
+  }
+
+  protected AbstractCastingRecipe findRecipe() {
+    if (this.recipe != null && this.recipe.matches(crafting, getWorld())) {
+      return this.recipe;
+    }
+    AbstractCastingRecipe castingRecipe = getWorld().getRecipeManager().getRecipe(this.recipeType, crafting, getWorld()).orElse(null);
+
+    return castingRecipe;
+  }
+
+  /** Called from CastingFluidHandler.fill()
+   * @param fluid   Fluid used in casting
+   * @param action  EXECUTE or SIMULATE
+   * @return        Amount of fluid needed for recipe, used to resize the tank.
+   */
+  public int initNewCasting(Fluid fluid, IFluidHandler.FluidAction action) {
+    this.crafting.setFluid(fluid);
+
+    AbstractCastingRecipe castingRecipe = findRecipe();
+    if (castingRecipe != null && this.recipe == null) {
```

Comment:
If you are going to skip if we have an existing recipe, might as well skip the call to `findRecipe` entirely. Can also return the size for the existing recipe in that case.